core/src/memory/binary_page_memory_resource.cpp

```diff
@@ -0,0 +1,348 @@
+                 * stack.
+                 */
+                rem.push(c->left.get());
+                rem.push(c->right.get());
+            }
+        }
+
+        /*
+         * If we have found the target, just mark it as vacant. There is no need
+         * to issue a deallocation upstream.
+         */
+        if (cand != nullptr) {
+            cand->free();
+        }
+    }
+
+    bool binary_page_memory_resource::do_is_equal(
+        const memory_resource & other
+    ) const noexcept {
+        /*
+         * All malloc resources are equal to each other, because they have no
+         * internal state. Of course they have a shared underlying state in the
+         * form of the underlying C library memory manager, but that is not
+         * relevant for us.
+         */
+        const binary_page_memory_resource * c;
+        c = dynamic_cast<const binary_page_memory_resource *>(&other);
+
+        return this == c;
+    }
+
+    binary_page_memory_resource::page * binary_page_memory_resource::find_free_page(
+        std::size_t size
+    ) {
+        page * cand = nullptr;
+
+        /*
+         * Here we also do a depth-first search, looking through all our pages
+         * to find the smallest free page that can fit our request.
+         */
+        std::stack<page *> rem;
+
+        for (std::unique_ptr<page> & p : m_pages) {
+            rem.push(p.get());
+        }
+
+        /*
+         * Core DFS loop.
+         */
+        while (!rem.empty()) {
+            page * c = rem.top();
+            rem.pop();
+
+            /*
+             * For split pages, we will need to examine both children as part of
+             * our search.
+             */
+            if (c->state == page_state::SPLIT) {
+                rem.push(c->left.get());
+                rem.push(c->right.get());
+            }
+
+            /*
+             * A page is a candidate if it's free, if it's big enough, and if
+             * it's smaller than any previous candidate.
+             */
+            if (c->is_free() && c->size >= size && (cand == nullptr || c->size < cand->size)) {
+                cand = c;
+            }
+
+            /*
+             * If the size of the page is exactly equal to the size of our
+             * rounded request, we will never find a smaller page and we can
+             * safely return what we have.
+             */
+            if (cand != nullptr && cand->size == size) {
+                break;
+            }
+        }
+
+        return cand;
+    }
+
+    void binary_page_memory_resource::allocate_upstream(
+        std::size_t size
+    ) {
+        /*
+         * Making too many small allocations here would be a bad idea, so we
+         * take a minimum size of 1 megabyte.
+         */
+        size = std::max(size, static_cast<std::size_t>(1048576));
+
+        /*
+         * Allocate the memory upstream and gather the address.
+         */
+        void * addr = m_upstream->allocate(size);
```

Comment:
I wonder if we should pass along the alignment value received by `do_allocate(...)`. We don't use the alignment for anything yet, but as long as [std::pmr::memory_resource](https://en.cppreference.com/w/cpp/memory/memory_resource) explicitly deals with memory alignment, our "high level" memory resources should also be aware of it.

Reply:
Yep, this is a good point.

Reply:
Thinking about it, due to the way this allocator works alignment is both a little harder and a little easier than normal... Passing on the argument here won't work as such, I think. Would you mind crossing that bridge when we get to it at a later date?